compose/core-compose/src/main/java/com/squareup/workflow/ui/compose/RenderAsState.kt

```diff
@@ -184,45 +182,65 @@ private class WorkflowState<PropsT, OutputT : Any, RenderingT>(
   initialProps: PropsT,
   private val outputRef: Ref<(OutputT) -> Unit>,
   private val snapshotState: MutableState<Snapshot?>,
-  private val diagnosticListener: WorkflowDiagnosticListener?
-) : CompositionLifecycleObserver {
+) : RememberObserver {
 
   private val renderingState = mutableStateOf<RenderingT?>(null)
-
-  // This can be a StateFlow once coroutines is upgraded to 1.3.6.
-  private val propsChannel = Channel<PropsT>(capacity = Channel.CONFLATED)
-    .apply { offer(initialProps) }
-  val propsFlow = propsChannel.consumeAsFlow()
-    .distinctUntilChanged()
+  private val propsFlow = MutableStateFlow(initialProps)
 
   // The value is guaranteed to be set before returning, so this cast is fine.
   @Suppress("UNCHECKED_CAST")
   val rendering: State<RenderingT>
     get() = renderingState as State<RenderingT>
 
   init {
-    launchWorkflowIn(workflowScope, workflow, propsFlow, snapshotState.value) { session ->
-      session.diagnosticListener = diagnosticListener
-
-      session.outputs.onEach { outputRef.value!!.invoke(it) }
-        .launchIn(this)
-
-      session.renderingsAndSnapshots
-        .onEach { (rendering, snapshot) ->
-          renderingState.value = rendering
-          snapshotState.value = snapshot
-        }
-        .launchIn(this)
+    val exitOutput = CompletableDeferred<OutputT>()
+    fun exit(code: OutputT) {
+      // If we don't cancel the workflow runtime explicitly, coroutineScope will hang waiting for it to
+      // finish.
+      workflowScope.coroutineContext.cancelChildren(
+          CancellationException("WorkflowState completed with exit code $exitOutput")
+      )
+      exitOutput.complete(code)
+    }
+
+    val renderings = renderWorkflowIn(
+        workflow = workflow,
+        scope = workflowScope,
+        props = propsFlow,
+        initialSnapshot = snapshotState.value?.bytes?.let { TreeSnapshot.parse(it) },
+        onOutput = { exit(it) }
+    )
+        .map { it.rendering }
+        .produceIn(workflowScope)
+
+    workflowScope.launch {
+      propsFlow.collect {
+        // TODO ask Zach
+      }
     }
+
+    // session.outputs.onEach { outputRef.value!!.invoke(it) }
+    //     .launchIn(this)
+    //
+    // session.renderingsAndSnapshots
+    //     .onEach { (rendering, snapshot) ->
+    //       renderingState.value = rendering
+    //       snapshotState.value = snapshot
+    //     }
+    //     .launchIn(this)
   }
 
   fun setProps(props: PropsT) {
-    propsChannel.offer(props)
+    propsFlow.value = props
   }
 
-  override fun onEnter() {}
+  override fun onRemembered() {}
+
+  override fun onForgotten() {
+    workflowScope.cancel()
+  }
 
-  override fun onLeave() {
+  override fun onAbandoned() {
     workflowScope.cancel()
   }
```

Comment:
Minor nit: I like to keep `onAbandoned` above `onRemembered` and `onForgotten` to hint that it happens earlier in the lifecycle.